File: minimumdepthofbinarytree.js
//Objective is to find the minimum depth of a binary tree.


//O(n) solution that uses BFS and a queue.

//Base case
if (!root) {
    return 0
}

let queue = []
let level = 1

queue.push(root)

while (queue.length > 0) {
    //This is to break the next while loop
    queue.push(null)

    //Explore the children
    while (queue[0] !== null) {
        let node = queue.shift()

        if (!node.left && !node.right) {
            return level
        }
        if (node.left) {
            queue.push(node.left)
        }

        if (node.right) {
            queue.push(node.right)
        }
    }

    //Get rid of the 'null' and move onto the next level
    queue.shift()
    level++
}

return level

File: minimumdepthofbinarytree.java
//Java Solution

class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        LinkedList<Pair<TreeNode, Integer>> stack = new LinkedList<>();
        stack.add(new Pair(root, 1));
        
        while (!stack.isEmpty()) {
            Pair<TreeNode, Integer> curr = stack.poll();
            TreeNode node = curr.getKey();
            int depth = curr.getValue();
            
            if ((node.left == null) && (node.right == null)) {
                return depth;
            }
            
            if (node.left != null) {
                stack.add(new Pair(node.left, depth + 1));
            }
            
            if (node.right != null) {
                stack.add(new Pair(node.right, depth + 1));
            }
        }
        
        return 0;
    }
}